nowcasting_dataset/dataset/datasets.py

```diff
@@ -338,3 +387,93 @@ def worker_init_fn(worker_id):
         # The NowcastingDataset copy in this worker process.
         dataset_obj = worker_info.dataset
         dataset_obj.per_worker_init(worker_info.id)
+
+
+def subselect_data(
+    batch: example.Example,
+    required_keys: Union[Tuple[str], List[str]],
+    current_timestep_index: int,
+    history_minutes: int,
+    forecast_minutes: int,
+) -> example.Example:
+    """
+    Subselects the data temporally. This function selects all data within the range [history_minutes, forecast_minutes]
+
+    Args:
+        batch: Example dictionary containing at least the required_keys
+        required_keys: The required keys present in the dictionary to use
+        current_timestep_index: The index into either SATELLITE_DATETIME_INDEX or NWP_TARGET_TIME giving the current timestep
+        history_minutes: How many minutes of history to use
+        forecast_minutes: How many minutes of future data to use for forecasting
+
+    Returns:
+        Example with only data between [history_minutes, forecast_minutes] remaining
+    """
+    # We are subsetting the data
+    date_time_index_to_use = (
+        SATELLITE_DATETIME_INDEX if SATELLITE_DATA in required_keys else NWP_TARGET_TIME
+    )
+    current_time = batch[date_time_index_to_use][0, current_timestep_index]
+    # Datetimes are in seconds, so just need to convert minutes to second + 30sec buffer
+    # Only need to do it for the first example in the batch, as masking indicies should be the same for all of them
+    start_time = current_time - (history_minutes * 60) - 30  # seconds
+    end_time = current_time + (forecast_minutes * 60) + 30  # seconds
+    if SATELLITE_DATA in required_keys and SATELLITE_DATETIME_INDEX in batch:
+        satellite_mask = np.logical_and(
```

Comment:
I could be wrong but I _think_ lines 422 to 430 could be replaced with something like:

```python
for key in [SATELLITE_DATA, SATELLITE_DATETIME_INDEX]:
    batch[key] = batch[key].sel(time=slice(start_time, end_time))
```

(But, to be honest, I haven't written any `xarray` code since late-July so I definitely could be wrong!)

(If this works, then the same principle could be applied to the code below)

Reply:
So, I did try doing something similar to that before, with a 2D array of start_times, end_times and it was giving errors, which is why I instead moved the subsetting code after the call to `example.to_numpy()`. I believe I might be able to do it with `.isel` if the indicies are all the same, but because each example in the batch has a different time period they cover, I don't think this would work?

Reply:
Aha, excellent point!  Sorry, I'm clearly not thinking straight!   Yeah, using `isel(time=slice(start_time, end_time))` feels like it should work (and might be less lines of code?)  But only if it makes sense to do the subsetting while the data is still in `xarray` objects

Reply:
(I wonder if we should get into the habit of using variable names for batches like `batch_np` and `batch_xr` to make it super-explicit when we're dealing with numpy vs xarray batches??  And, yeah, I'd maybe lean towards converting to numpy as the very last step (and only because PyTorch expects numpy arrays, and doesn't know what to do with xarray objects!)

Reply:
It turns out that doing that doesn't work unfortunately, so not sure it'll actually be more concise, but I do agree doing it all in xarray is probably cleaner. I tried it this way: 
`current_time = batch[date_time_index_to_use].isel(time=current_timestep_index)
    start_time = current_time - pd.to_timedelta(f"{history_minutes} minute 30 second")
    end_time = current_time + pd.to_timedelta(f"{forecast_minutes} minute 30 second")
    subselect = batch[SATELLITE_DATA].isel(time=(start_time.data, end_time.data))`

Reply:
Got it working, and updated the code! Just need to change the test now.

Reply:
Ah, got it... Sorry to be boring about this but please can we have one more try :)

As you probably know, `isel` expects _indicies_ into the array (just as if you were indexing into a numpy array).  For example, `isel(time=slice(4, 6))` on an xarray object is just like `array[4:6]` on a numpy array. 

So, to use `isel`, the code must first find the indicies into `time` at which `start_time` and `end_time` appear.  Finding the indicies can be done with [`np.searchsorted(a, v)`](https://numpy.org/doc/stable/reference/generated/numpy.searchsorted.html) (which finds the index in `a` at which you'd insert `v` into `a` such that `a` remains sorted).

So, something vaguely like this might work (although it might be off by one!)

```python
    start_time = current_time - pd.to_timedelta(f"{history_minutes} minute 30 second")
    end_time = current_time + pd.to_timedelta(f"{forecast_minutes} minute 30 second")
    time_index_of_first_batch = batch[SATELLITE_DATA].time[0]
    start_i, end_i = np.searchsorted(time_index_of_first_batch, [start_time, end_time])
    subselect = batch[SATELLITE_DATA].isel(time=slice(start_i, end_i))
```

Reply:
Okay! That's a bit simpler than what I have working now, so I'll try that out!

Reply:
Nice!  (Oops, sorry, I think we were both figuring this out at the same time - I don't think your message at 18:31 had appeared when I was writing my reply!  But well done for getting it working without my ramblings!)

Reply:
Had to make a few changes, but it works and is a bit more concise than what I had. The only issue left is getting the xarray for the unit test setup.

Reply:
Its all working now!